_posts/2022-01-10-bevy-labels.md

```diff
@@ -0,0 +1,214 @@
+Look at all these bounds!
+You might recognize a few from usual Rust code,
+but `DynHash` stands out as one trait defined in `bevy::utils`.
+We'll come back to it later, and just treat it as the regular `Hash` trait for now.
+
+`SystemLabel` also looks like an empty trait -- but that's actually an illusion.
+Its only item is hidden in the docs.
+We can assume that's because its an implementation detail,
+and instead of implementing this trait manually,
+we are supposed to derive it.
+Indeed, there is a [`SystemLabel` derive macro].
+
+Okay, so to get a type to be a `SystemLabel`,
+it needs to implement `Debug`, `Hash`
+(the compiler will figure `'static + Send + Sync` out for us).
+As you might know, to derive `Hash`, we also need to derive `PartialEq + Eq`.
+And by experimentation and reading compiler errors,
+we can see that the `SystemLabel` derive actually also adds a requirement on `Clone`.
+
+In the end we arrive at something like this:
+
+```rust
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemLabel)]
+enum System {
+    Input,
+}
+```
+
+Which we can use just like our string previously:
+
+```rust
+.add_startup_system(mouse_input.label(System::Input))
+.add_startup_system(move_player.after(System::Input))
+```
+
+[`SystemLabel` derive macro]: https://docs.rs/bevy/0.6.0/bevy/ecs/schedule/derive.SystemLabel.html
+
+## Some notes on the magic
+
+So what is the deal with that [`DynHash`] trait?
+If you look at the [API docs][`DynHash`],
+you can see that it requires an implementation of `DynEq`
+(also from Bevy),
+which in turn requires an implementation of `Any`.
+
+Its methods are also kind of strange:
+Compared to standard library's [`Hash`] trait,
+there are no generics, but a lot of `dyn` keywords.
+This looks to me that someone went out of their way to make
+an [object-safe] version of `Hash`.
+
+The good news is: Users of the API don't have to care:
+`DynHash` is implemented for all types that implement `Hash` and `DynEq`,
+and `DynEq` is in turn implemented for all types that implement `Eq` and `Any`,
+
+Another thing that seems magical is that hidden trait method on `SystemLabel`,
+which is actually called `dyn_clone`.
+Similar to the other dynamic trait implementations,
+this allows cloning any `SystemLabel` type,
+even if all you have is a `Box<dyn SystemLabel>`.
+
+
+[`DynHash`]: https://docs.rs/bevy/0.6.0/bevy/utils/label/trait.DynHash.html
+[`Hash`]: https://doc.rust-lang.org/1.57.0/core/hash/trait.Hash.html
+[object-safe]: https://doc.rust-lang.org/book/ch17-02-trait-objects.html
+
+## Generic label types
+
+Did you think we were done?
+Oh no! There one more thing:
+`SystemLabel` is not alone!
+There is also `StageLabel`, `AmbiguitySetLabel`, and `RunCriteriaLabel`.
+
+These label types are pretty much all the same,
+but distinct traits in the type system.
+That means you will have to explicitly derive `StageLabel`
+if you want to use your type to refer to a stage;
+you can't use a `SystemLabel` or any other label for that.
```

Comment:
This is an important point that I think deserves a bit more clarification. The fact that you can't use system labels for stage labels is very helpful: it prevents you from feeding the compiler garbage instructions, and moves those errors from run time to compile time.

Reply:
Good point! Will add a paragraph :)